### PR TITLE
add the small perimeter length parameter

### PR DESCRIPTION
**Pull request by gennartan**
Implements #4294

### CODE REVIEW

**Comment by lordofhyphens** [comment]
A lot of the "changes" are whitespace changes. Please revert those changes and update the PR. 

If you want to commit a separate PR for code beautification, go ahead, but please don't mix functional changes and completely unrelated aesthetic changes.

**Comment by lordofhyphens** [comment]
@gennartan Please provide a general description of a test for this feature; as it stands, there's no way to tell if this code breaks in the future (go ahead and add it just to src/test/libslic3r).

File: src/test/inputs/test_config/config_bundle_test.ini [hunk 1]
```diff
@@ -74,6 +74,7 @@ skirt_distance = 6
 skirt_height = 1
 skirts = 1
 small_perimeter_speed = 15
+small_perimeter_length = 20
```
Comment (VanessaE) [inline src/test/inputs/test_config/config_bundle_test.ini]: Maybe make this a bit larger by default?  20 mm would translate to a hole bit under 6.4 mm diameter.  I would suggest a default of 32, for a hole about 1 cm diameter.

File: xs/src/libslic3r/GCode.cpp [hunk 2]
```diff
@@ -391,23 +391,23 @@ GCode::extrude(ExtrusionLoop loop, std::string description, double speed)
...
     // get paths
     ExtrusionPaths paths;
     loop.clip_end(clip_length, &paths);
     if (paths.empty()) return "";
-    
+
     // apply the small perimeter speed
     if (paths.front().is_perimeter()
         && !loop.has(erOverhangPerimeter)
-        && loop.length() <= SMALL_PERIMETER_LENGTH
```
Comment (VanessaE) [inline xs/src/libslic3r/GCode.cpp]: I think you forgot to remove this constant's declaration?

https://github.com/slic3r/Slic3r/blob/fb2fefd4daa5600b7b46b428ba4278f3dc5a6e11/xs/src/libslic3r/libslic3r.h#L86

**Comment by lordofhyphens** [comment]
@gennartan probably the "easiest" way to revert these changes is through the use of gitk and git-gui.

1. Start gitk and find the commit before you started to make your changes.
Right-click on the commit and select "reset branch to here"
image
1. Select "mixed" and click "OK"
image
1. With git-gui, select just the functional changes, and revert the rest.
image
1. Rewrite your commit message. 
1. Commit.
1. `git push --force` from the command line. 

This gets the diff of the changes down to around 15 lines of code (which is about what you'd expect from this change) and makes it much easier to tell what is going on.

In the future, using a different branch than master on your fork will cause fewer problems as well on your end (more flexibility), as more commits to master on your system will cause the PR to get updated.

[changes.txt](https://github.com/slic3r/Slic3r/files/2568964/changes.txt)

File: xs/src/libslic3r/PrintConfig.cpp [hunk 3]
```diff
@@ -1281,6 +1281,15 @@ PrintConfigDef::PrintConfigDef()
...
     def->default_value = new ConfigOptionFloatOrPercent(15, false);
 
+    def = this->add("small_perimeter_length", coFloat);
+    def->label = __TRANS("Small perimeter length");
+    def->category = __TRANS("Layers and Perimeters");
+    def->tooltip = __TRANS("This setting controls the length of the small perimeter parameter.");
+    def->sidetext = "mm";
+    def->cli = "small-perimeter-length=f";
+    def->min = 0;
+    def->default_value = new ConfigOptionFloat(20);
```
Comment (lordofhyphens) [inline xs/src/libslic3r/PrintConfig.cpp]: set the default to `6.5 * 2 * PI`, so that we aren't breaking existing configurations.

File: src/test/inputs/test_config/config_bundle_test.ini [hunk 4]
```diff
@@ -74,7 +74,7 @@ skirt_distance = 6
 skirt_height = 1
 skirts = 1
 small_perimeter_speed = 15
-small_perimeter_length = 20
+small_perimeter_length = 40
```
Comment (lordofhyphens) [inline src/test/inputs/test_config/config_bundle_test.ini]: Why is this changed here? It is an input for a config test.
